multidict/_multidict_py.py

```diff
@@ -455,15 +454,204 @@ def _identity(self, key: str) -> str:
             raise TypeError("MultiDict keys should be either str or subclasses of str")
 
 
+def estimate_log2_keysize(n: int) -> int:
+    # 7 == HT_MINSIZE - 1
+    return (((n * 3 + 1) // 2) | 7).bit_length()
+
+
+@dataclass
+class _Entry(Generic[_V]):
+    hash: int
+    identity: str
+    key: str
+    value: _V
+
+
+@dataclass
+class _HtKeys(Generic[_V]):  # type: ignore[misc]
+    LOG_MINSIZE: ClassVar[int] = 3
+    MINSIZE: ClassVar[int] = 8
+    PREALLOCATED_INDICES: ClassVar[dict[int, array]] = {  # type: ignore[type-arg]
+        log2_size: array(
+            "b" if log2_size < 8 else "h", (-1 for i in range(1 << log2_size))
+        )
+        for log2_size in range(3, 10)
+    }
+
+    log2_size: int
+    usable: int
+
+    indices: array  # type: ignore[type-arg] # in py3.9 array is not a generic
+    entries: list[Optional[_Entry[_V]]]
+
+    @functools.cached_property
+    def nslots(self) -> int:
+        return 1 << self.log2_size
+
+    @functools.cached_property
+    def mask(self) -> int:
+        return self.nslots - 1
+
+    if sys.implementation.name != "pypy":
+
+        def __sizeof__(self) -> int:
+            return (
+                object.__sizeof__(self)
+                + sys.getsizeof(self.indices)
+                + sys.getsizeof(self.entries)
+            )
+
+    @classmethod
+    def new(cls, log2_size: int, entries: list[Optional[_Entry[_V]]]) -> Self:
+        size = 1 << log2_size
+        usable = (size << 1) // 3
+        if log2_size < 10:
+            indices = cls.PREALLOCATED_INDICES[log2_size].__copy__()
+        elif log2_size < 16:
+            indices = array("h", (-1 for i in range(size)))
+        elif log2_size < 32:
+            indices = array("l", (-1 for i in range(size)))
+        else:  # pragma: no cover  # don't test huge multidicts
```

Comment:
Why not?

Reply:
Because of memory: a multidict with 2^32 items takes more than 128 GiB, it is much more than the memory size of the test box.

Reply:
Is it cheatable? e.g. If we used a log2_size==32, but actually only passed 1 or 2 items. Or is the empty allocated array already too big?

Reply:
I meant cheat the test, not change the function. But, sounds like it's not going to work.